simexpal/base.py

```diff
@@ -693,72 +693,191 @@ def check_available(self):
+
+			elif 'generator' in self._inst_yml:
+				import subprocess
+
+				def substitute(p):
+					if p == 'INSTANCE_FILENAME':
+						return self.unique_filename
+					raise RuntimeError("Unexpected parameter {}".format(p))
+
+				print("Generating instance '{}'".format(self.shortname))
+
+				assert isinstance(self._inst_yml['generator']['args'], list)
+				cmd = [util.expand_at_params(arg_tmpl, substitute) for arg_tmpl
+						in self._inst_yml['generator']['args']]
+
+				partial_path = os.path.join(instance_dir, self.unique_filename)
+				with open(partial_path + '.gen', 'w') as f:
+					subprocess.check_call(cmd, cwd=self.config.basedir,
+							stdout=f, stderr=subprocess.PIPE)
+				os.rename(partial_path + '.gen', partial_path + '.post0')
+			elif 'method' in self._inst_yml:
+				print(f"Downloading instance '{self.shortname}' with method '{self.method}'")
+
+				instances.download_instance(self._inst_yml,	instance_dir, self.unique_filename,
+					os.path.join(instance_dir, self.unique_filename), '.post0')
+			else:
+				raise RuntimeError(f"Unknown installation option for instance '{self.shortname}'")
 
 		stage = 0
+		filenames = self.filenames
 		if 'postprocess' in self._inst_yml:
+
 			did_work = False
-			if not os.path.isfile(partial_path + '.postprocessed'):
-				assert self._inst_yml['postprocess'] == 'to_edgelist'
-				instances.convert_to_edgelist(self._inst_yml,
-						partial_path + '.post0', partial_path + '.post1')
-				os.rename(partial_path + '.post0', partial_path + '.original')
-				stage = 1
-				did_work = True
-				util.touch(partial_path + '.postprocessed')
+			postprocessed = os.path.isfile(os.path.join(instance_dir, self.yml_name) + '.postprocessed')
+			if not postprocessed:
+
+				# Make sure that the .post0 files exist.
+				# For cases like 'repo: local' instances, adding the 'postprocess' key
+				# after downloading the instance or re-postprocessing instances.
+				for file in filenames:
+					if not os.path.isfile(os.path.join(instance_dir, file + '.post0')):
+						os.rename(os.path.join(instance_dir, file), os.path.join(instance_dir, file + '.post0'))
+
+				print("simexpal: Start postprocessing of instance '{}'".format(self.shortname))
+				if isinstance(self._inst_yml['postprocess'], str):
+					assert self._inst_yml['postprocess'] == 'to_edgelist'
+
+					partial_path = os.path.join(instance_dir, self.unique_filename)
+					try:
+						print("simexpal: Converting instance '{}' to edge list format".format(self.shortname))
+						instances.convert_to_edgelist(self._inst_yml, partial_path + '.post0', partial_path + '.post1')
+						os.rename(partial_path + '.post0', partial_path + '.original')
+					except RuntimeError as e:
+						print("simexpal: An error occurred while postprocessing the instance '{}': {}".format(
+							self.shortname, e))
+						print("simexpal: Restoring original instance files")
+						os.rename(partial_path + '.post0', partial_path)
```

Comment:
Hmmm, if we restore the instance file to the `post0` file, how do we know if any given instance has already been postprocessed or not?

Reply:
The old logic here was: the `post0` files are obtained through download, and the instance files are *always* already postprocessed, right?

Reply:
> Hmmm, if we restore the instance file to the `post0` file, how do we know if any given instance has already been postprocessed or not?

For each instance we create a ``.postprocessed`` file (lines ``781`` and ``868``) when the postprocessing was successful. Also, we print out status messages in the terminal. Although we probably shouldn't assume that it is enough. Maybe I could add an additional check in ``Instance.check_available()`` which looks for the ``.postprocessed`` file for instances that are supposed to be postprocessed. In this way, experiments can't be started on unprocessed instances that were supposed to be postprocessed.

> The old logic here was: the `post0` files are obtained through download, and the instance files are _always_ already postprocessed, right?

Yes. With this PR ``.post0`` files are still obtained through download and instance files are postprocessed depending on the existence of the ``.postprocessed`` file

Reply:
Concerning `.postprocessed` and `Instance.check_available()`: I think it is enough to inform the user when postprocessing has not been successful. Making successful postprocessing mandatory might lead to unexpected results for users (for example if the experiment does not really rely on it in practice).